### PR TITLE
Add Store comonad

### PR DESCRIPTION
**Pull request by rilut**
Hi, I'm trying to implement the Store comonad (#100), but I guess I need to `extends` FantasyComonad and StaticComonad? Also what to do with `F extends HKTS`?

Thanks,
rilut

### CODE REVIEW

**Comment by gcanti** [comment]
Hi, 

The general process to define new data structures in `fp-ts` follows the steps below (I'm using [this](https://github.com/purescript/purescript-transformers/blob/master/src/Control/Comonad/Store/Class.purs) as reference implementation (PureScript))

(1) define the new data structure

```ts
export class Store<S, A> {
  constructor(
    public readonly peek: (s: S) => A,
    public readonly s: S
  ) {}
}
```

(2) make the new data structure a higher kinded type

```ts
declare module 'fp-ts/lib/HKT' {
  interface HKT<A, U> {
    Store: Store<U, A> // <= important: the key 'Store' here must correspond to the value of URI defined below, this makes 'Store' a member of HKTS
  }
}
```

(3) implement fantasy type classes, in this case `FantasyComonad`

```ts
export const URI = 'Store' // <= important: must be a unique identifier

export type URI = typeof URI

export class Store<S, A> implements FantasyComonad<URI, A> {
  constructor(
    public readonly peek: (s: S) => A,
    public readonly s: S
  ) {}
  pos(): S {
    return this.s
  }
  map<B>(f: (a: A) => B): Store<S, B> {
    return new Store(s => f(this.peek(s)), this.s)
  }
  extract(): A {
    return this.peek(this.s)
  }
  extend<B>(f: (wa: Store<S, A>) => B): Store<S, B> {
    return new Store<S, B>(s => f(this), this.s)
  }
}
```

(4) implement static type classes, in this case `StaticComonad`

```ts
export function map<S, A, B>(f: (a: A) => B, wa: Store<S, A>): Store<S, B> {
  return wa.map(f)
}

export function extract<S, A>(wa: Store<S, A>): A {
  return wa.extract()
}

export function extend<S, A, B>(f: (wa: Store<S, A>) => B, wa: Store<S, A>): Store<S, B> {
  return wa.extend(f)
}
```

(5) add other useful functions

```ts
/** reads the current position */
export function pos<S, A>(wa: Store<S, A>): S {
  return wa.pos()
}

/** reads the value at the specified position in the specified context */
export function peek<S, A>(wa: Store<S, A>, s: S): A {
  return wa.peek(s)
}

/** Extract a value from a position which depends on the current position */
export function peeks<S, A>(f: Endomorphism<S>, wa: Store<S, A>, s: S): A {
  return wa.peek(f(wa.pos()))
}

/** Reposition the focus at the specified position */
export function seek<S, A>(s: S, wa: Store<S, A>): Store<S, A> {
  return new Store(wa.peek, s)
}

/** Reposition the focus at the specified position, which depends on the current position */
export function seeks<S, A>(f: Endomorphism<S>, wa: Store<S, A>): Store<S, A> {
  return new Store(wa.peek, f(wa.pos()))
}

/** Extract a collection of values from positions which depend on the current position */
export function experiment<F extends HKTS, S, A>(functor: StaticFunctor<F>, f: (s: S) => HKT<S>[F], wa: Store<S, A>): HKT<A>[F] {
  return functor.map<S, A>(s => wa.peek(s), f(wa.pos()))
}
```

Note: a default `duplicate` function is already provided by the `Extend` module 

---

Result (not tested)

```ts
import { HKT, HKTS } from 'fp-ts/lib/HKT'
import { StaticComonad, FantasyComonad } from 'fp-ts/lib/Comonad'
import { StaticFunctor } from 'fp-ts/lib/Functor'
import { Endomorphism } from 'fp-ts/lib/function'

export const URI = 'Store'

export type URI = typeof URI

declare module 'fp-ts/lib/HKT' {
  interface HKT<A, U> {
    Store: Store<U, A>
  }
}

export class Store<S, A> implements FantasyComonad<URI, A> {
  constructor(
    public readonly peek: (s: S) => A,
    public readonly s: S
  ) {}
  pos(): S {
    return this.s
  }
  map<B>(f: (a: A) => B): Store<S, B> {
    return new Store(s => f(this.peek(s)), this.s)
  }
  extract(): A {
    return this.peek(this.s)
  }
  extend<B>(f: (wa: Store<S, A>) => B): Store<S, B> {
    return new Store<S, B>(s => f(this), this.s)
  }
}

export function map<S, A, B>(f: (a: A) => B, wa: Store<S, A>): Store<S, B> {
  return wa.map(f)
}

export function extract<S, A>(wa: Store<S, A>): A {
  return wa.extract()
}

export function extend<S, A, B>(f: (wa: Store<S, A>) => B, wa: Store<S, A>): Store<S, B> {
  return wa.extend(f)
}

/** reads the current position */
export function pos<S, A>(wa: Store<S, A>): S {
  return wa.pos()
}

/** reads the value at the specified position in the specified context */
export function peek<S, A>(wa: Store<S, A>, s: S): A {
  return wa.peek(s)
}

/** Extract a value from a position which depends on the current position */
export function peeks<S, A>(f: Endomorphism<S>, wa: Store<S, A>, s: S): A {
  return wa.peek(f(wa.pos()))
}

/** Reposition the focus at the specified position */
export function seek<S, A>(s: S, wa: Store<S, A>): Store<S, A> {
  return new Store(wa.peek, s)
}

/** Reposition the focus at the specified position, which depends on the current position */
export function seeks<S, A>(f: Endomorphism<S>, wa: Store<S, A>): Store<S, A> {
  return new Store(wa.peek, f(wa.pos()))
}

/** Extract a collection of values from positions which depend on the current position */
export function experiment<F extends HKTS, S, A>(functor: StaticFunctor<F>, f: (s: S) => HKT<S>[F], wa: Store<S, A>): HKT<A>[F] {
  return functor.map<S, A>(s => wa.peek(s), f(wa.pos()))
}
```

**Comment by gcanti** [comment]
Opss.. this is the new way (v0.3)

```ts
declare module 'fp-ts/lib/HKT' {
  interface HKT<A, U> {
    Store: Store<U, A> // <= important: the key 'Store' here must correspond to the value of URI defined below, this makes 'Store' a member of HKTS
  }
}
```

The old way (v0.2.x) is

```ts
declare module './HKT' {
  interface HKT<A> {
    Store: Store<any, A>
  }
  interface HKT2<A, B> {
    Store: Store<A, B>
  }
}
```

**Comment by gcanti** [comment]
In addition, step (6), you can add the following phantom types to improve type safety

```ts
readonly _S: S
readonly _A: A
readonly _URI: URI
```

This typically prevents to mix wrong instances

**Comment by rilut** [comment]
oops, spacing issue. let me configure my editor first

File: src/Store.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+/** Extract a collection of values from positions which depend on the current position */
+export function experiment<F extends HKTS, S, A>(functor: StaticFunctor<F>, f: (s: S) => HKT<S>[F], sa: Store<S, A>): HKT<A>[F] {
+  return functor.map<S, A>(s => sa.peek(s), f(sa.pos()))
+}
+
+// tslint:disable-next-line no-unused-expression
+;(
+  { map, extract, extend } as (
+    StaticComonad<URI> &
+    StaticFunctor<URI>
```
Comment (gcanti) [inline src/Store.ts]: FYI: no need to check `Functor` here, checking `Comonad` is enough (`Comonad` extends `Extends` which in turn extends `Functor`)

Comment (rilut) [inline src/Store.ts]: @gcanti updated. thanks

Comment (rilut) [inline src/Store.ts]: btw that's because what I've seen on Tuple.ts, so can I remove the `StaticFunctor` on Tuple.ts on different commit?

Comment (gcanti) [inline src/Store.ts]: ah, you are right, thanks for pointing out

File: src/Store.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  return new Store(sa.peek, s)
+}
+
+/** Reposition the focus at the specified position, which depends on the current position */
+export function seeks<S, A>(f: Endomorphism<S>, sa: Store<S, A>): Store<S, A> {
+  return new Store(sa.peek, f(sa.pos()))
+}
+
+/** Extract a collection of values from positions which depend on the current position */
+export function experiment<F extends HKTS, S, A>(functor: StaticFunctor<F>, f: (s: S) => HKT<S>[F], sa: Store<S, A>): HKT<A>[F] {
```
Comment (rilut) [inline src/Store.ts]: sure

**Comment by gcanti** [comment]
Thanks @rilut 👍

**Comment by rilut** [comment]
Thank you @gcanti, for being very helpful 😄

**Pull request merged**